[Fia] step-1 / step-2 구현

yeonise opened the pull request:
안녕하세요, Cooper. Fia입니다.

## 전체적인 학습 과정

Spring에 대한 학습이나 경험이 부족하기 때문에,
우선 미션을 진행하면서 필요한 부분을 찾아 코드를 작성하고 프로그램을 동작해 보면서
어떤 식으로 사용하는지 경험해 보고자 했습니다.  

학습 계획은 전날 진행 상황을 보고 아침마다 작성하였습니다.

## 이번 주 진행 상황

1. Spring MVC 강의를 듣기 시작했습니다. 이번 주에 강의를 통해 학습한 내용은 다음과 같습니다.
  - HttpServletRequest, HttpServletResponse, HttpServlet의 service()를 사용하여 요청을 주고 받기
  - Servlet과 JSP 사용해보기
  - MVC 패턴 개요
2. mission step-1과 step-2를 구현했습니다.
  - 팀원들과 함께 진행한 그룹 리뷰의 피드백과 MVC 강의 뒷 부분을 찾아보면서 진행했습니다.
  - 왜 이렇게 동작하는지, 어떻게 돌아가는지, 왜 필요한지에 대해서 깊게 학습하지 못했습니다.

## 구현 내용

### Step-1

1. 회원 가입 기능 구현
2. 회원 목록 조회 기능 구현
3. 회원 프로필 조회 기능 구현
4. HTML 중복 제거

### Step-2

1. 글 작성 기능 구현
2. 글 목록 조회 기능 구현
3. 글 상세 보기 기능 구현
4. 회원 정보 수정 기능 구현

### URL

| HTTP | URL | 기능 |
| --- | --- | --- |
| `GET` | /users/join | 회원 가입 페이지 |
| `POST` | /users/join | 회원 가입 |
| `GET` | /users | 회원 목록 조회 |
| `GET` | /users/{userId} | 회원 프로필 정보 보기 |
| `GET` | /users/{userId}/update | 회원 정보 수정 페이지 |
| `POST` | /users/{userId}/update | 회원 정보 수정 |
| `GET` | /articles/write | 글 작성 페이지 |
| `POST` | /articles/write | 글 작성 |
| `GET` | / | 글 목록 조회 |
| `GET` | /articles/{id} | 글 상세 보기 |

### 사용해 본 것

- spring-boot-starter-web
- thymeleaf
- thymeleaf-layout을 사용하여 중복된 html 정리하기
- html, css, js 사용하여 front 단에서 유효성 검사하기
- controller - repository - domain으로 분리
- http request 안의 model에 데이터를 담아서 넘기기
- redirect
- interface 분리

### 실행 화면

**회원 가입**  
SpringBoot에서 제공하는 유효성 검사 없이,  
아이디 중복을 확인하여 사용자에게 안내하는 방법을 찾아내지 못했습니다.  
정규식을 사용하여 JS로 front 단에서 형식만 검증합니다.

<img width="800" alt="스크린샷 2023-03-31 오후 2 59 56" src="https://user-images.githubusercontent.com/105152276/229035846-ebbdfa92-0678-4744-893e-c989fb3facb8.png">

**회원 목록 조회**  
아직 html 파일을 만들지 못해서 샘플 파일을 사용하였습니다.

<img width="800" alt="스크린샷 2023-03-31 오후 3 12 28" src="https://user-images.githubusercontent.com/105152276/229038008-2eb53fe1-e939-4041-9c89-3aa14e26dc22.png">

**글 작성**  
글 작성 또한 정규식을 사용하여 JS로 front 단에서 형식만 검증합니다.

<img width="800" alt="스크린샷 2023-03-31 오후 3 16 22" src="https://user-images.githubusercontent.com/105152276/229038659-dd225d67-2d7a-4b62-a875-4fd5bb217b26.png">

**글 목록 조회**  
글 제목을 클릭하면 상세 보기 페이지로 이동합니다.

<img width="800" alt="스크린샷 2023-03-31 오후 3 19 56" src="https://user-images.githubusercontent.com/105152276/229039275-0c60aa0a-28f5-4095-8b6b-8e065fe922d5.png">

**글 상세 보기**  
댓글 기능은 아직 구현하지 않았습니다.

<img width="800" alt="스크린샷 2023-03-31 오후 3 25 58" src="https://user-images.githubusercontent.com/105152276/229040524-45febebf-0cff-4e27-850b-b76703364992.png">

## 고민

commit의 단위나 message 작성 관련하여 조언을 들을 수 있을까요?
어디까지 작성하고 나눠야 할지, convention을 따르는 게 좋을지 고민입니다.

읽어주셔서 감사합니다 : )

yeonise committed:
3월 27일 월요일 학습 계획 작성

README.md Update

yeonise committed:
static 폴더에서 templates 폴더로 이동 후 Thymeleaf 적용

user/form.html과 user/list.html을 templates의 하위 폴더인 user로 이동합니다.
index.html을 templates로 이동합니다.
---
각 파일에 thymeleaf 경로를 추가합니다.
구현할 페이지의 a href 속성을 th:href로 변경합니다.

yeonise committed:
User Domain 생성

아이디, 비밀번호, 이름, 이메일을 갖는 User Domain을 생성합니다.
---
getter와 setter를 추가합니다.

yeonise committed:
User Controller 생성

URL Convention은 다음과 같습니다.
- 회원 가입
GET : /users/join
POST : /users/join
- 회원 목록 조회
GET : /users
- 회원 프로필 정보 보기
GET : /users/{userId}

yeonise committed:
User Repository 생성

JCF - Map을 사용하여 데이터를 임시 저장합니다.
---
clearStore()는 테스트 시 사용합니다.

yeonise committed:
User Repository Test 생성

User Repository의 각 메서드가 예상대로 동작하는지 확인합니다.
---
테스트를 통과한 메서드는 다음과 같습니다.
- save()
- findByEmail()
- findAll()

yeonise committed:
fix : User Domain의 getter 수정

userId의 getter를 getId()에서 getUserId()로 수정합니다.

yeonise committed:
User Controller에 회원 목록을 처리하는 Get Mapping 추가

Model에 모든 회원 목록을 추가합니다.
---
/users : 모든 회원 목록 조회 페이지입니다.

yeonise committed:
회원 조회 페이지의 User List 구현

Repository에 저장된 회원 목록을 th:each를 사용하여 list.html에 추가합니다.

yeonise committed:
3월 28일 화요일 학습 계획 작성

README.md Update

yeonise committed:
3월 29일 수요일 학습 계획 작성

README.md Update

yeonise committed:
ID 값을 email에서 userId로 변경

email을 ID 값으로 사용하던 것을 userId로 변경합니다.
따라서 프로필 페이지의 요청 파라미터에 email 대신 userId를 사용합니다.

yeonise committed:
회원 조회 페이지에서 사용자 이름에 a 태그 추가

회원 목록 페이지에서 회원의 이름을 클릭하면 해당하는 회원의 프로필 페이지로 이동합니다.
요청 파라미터로 userId를 사용합니다.
---
list.html의 navbar에서 불필요한 active를 제거합니다.

yeonise committed:
User Controller에 회원 프로필을 처리하는 Get Mapping 추가

@PathVariable을 통해 요청 파라미터를 받아 사용합니다.
Model에 담아서 넘기면 View Template Engine이 받아서 화면에 처리합니다.

yeonise committed:
profile.html 파일을 static 폴더에서 templates 폴더로 이동 후 Thymeleaf 적용

yeonise committed:
Thymeleaf Layout Dependency 추가

build.gradle에 Thymeleaf Layout Dependency를 추가합니다.

yeonise committed:
head 태그의 공통된 설정을 config.html로 분리

yeonise committed:
header 부분을 header.html로 분리

페이지 상단에 고정되어 반복되는 header를 분리합니다.

yeonise committed:
공통된 script 파일들을 scripts.html로 분리

yeonise committed:
Default Layout을 생성

공통된 config, header, scripts를 갖는 기본 레이아웃을 생성합니다.
각 페이지의 내용은 content block 안에 작성합니다.

yeonise committed:
index.html에 layout을 적용하여 중복 제거

yeonise committed:
form.html에 layout을 적용하여 중복 제거

yeonise committed:
list.html에 layout을 적용하여 중복 제거

yeonise committed:
profile.html에 layout을 적용하여 중복 제거

yeonise committed:
fix: head 태그의 title이 적용되지 않는 오류 수정

head 태그 자체를 추가하는 방식에서
head 태그 안에 config 파일을 추가하는 방식으로 수정합니다.
이제 각 페이지마다 title을 정할 수 있습니다.

yeonise committed:
각 페이지의 title 태그 수정

yeonise committed:
Step-1 구현 완료

README.md Update

yeonise committed:
3월 30일 목요일 학습 계획 작성

README.md Update

yeonise committed:
User Repository Interface 생성

기존의 UserRepository Class를 Interface를 구현한 MemoryUserInterface로 수정합니다.
---
User Domian에 Long id를 추가합니다.

yeonise committed:
Article Domain 생성

인덱스, 제목, 작성자, 내용, 등록일을 갖는 Article Domain을 생성합니다.
---
getter와 setter를 추가합니다.

yeonise committed:
Article Repository Interface 생성

ArticleRepository Interface를 구현한 MemoryArticleRepository도 추가합니다.
---
MemoryArticleRepositoryTest도 추가합니다.

yeonise committed:
Article Controller 생성

URL은 다음과 같습니다.
- 글 작성
GET : /articles/write
POST : /articles/write
- 글 목록 조회
GET : /
- 글 상세 보기
GET : /articles/{id}

yeonise committed:
static 폴더에서 templates 폴더로 이동 후 Thymeleaf 적용

이동한 파일은 다음과 같습니다.
- form.html
- show.html

yeonise committed:
불필요한 파일 삭제

static의 /user/list.html을 삭제합니다.

yeonise committed:
/user/form.html의 Thymeleaf 문법 오류 수정

th:action에 경로 미지정 오류를 수정합니다.

yeonise committed:
content 뒤에 's'가 붙은 문법 오류 수정

Rename 과정에서 content 뒤에 's'가 붙어 contents로 변경된 문법 오류를 수정합니다.

yeonise committed:
게시글 목록 조회 페이지 구현

회원 목록 조회 페이지를 참고하여 게시글 목록 조회 페이지를 구현합니다.
---
table style 속성도 추가합니다.

yeonise committed:
Script 경로 수정

잘못된 상대 경로에서 절대 경로로 수정합니다.

yeonise committed:
불필요한 header 중복 제거

불필요한 th:replace를 제거합니다.

yeonise committed:
회원 가입 Front 단에서 유효성 검증 구현

전에 만들었던 회원 가입 페이지를 사용하여 Front 단에서 유효성을 검증합니다.

yeonise committed:
UserRepository에 isExists(String userId) 추가

입력한 userId가 이미 존재하는지 확인하여 boolean 값을 반환합니다.

yeonise committed:
글 작성 페이지의 HTML 파일 변경

기존에 만들었던 HTML CSS JS 파일을 사용하여 Front 단에서 유효성을 검사합니다.
---
글 작성자를 입력했는지 확인하는 checkWriter() 메서드를 추가합니다.

yeonise committed:
게시글 목록 조회 페이지의 HTML 파일 변경

메인 페이지의 HTML 파일을 변경합니다.
---
글 목록 조회 페이지를 메인 페이지로 사용합니다.

yeonise committed:
qna에서 article로 패키지 변경

qna에서 article 패키지로 이동한 파일은 다음과 같습니다.
- list.html
- show.html

yeonise committed:
글 작성 페이지 CSS 파일 추가

yeonise committed:
3월 31일 금요일 학습 계획 작성

README.md Update

yeonise committed:
회원 정보 수정 페이지 생성

기존의 아이디, 이메일, 닉네임은 미리 작성하여 보여줍니다.

yeonise committed:
회원 정보 수정을 위한 유효성 검사 추가

수정한 이메일, 닉네임, 비밀번호의 형식을 검사합니다.
아이디는 수정할 수 없습니다.

yeonise committed:
회원 정보 수정 페이지 연결

회원 목록 오른쪽에 위치한 수정 버튼을 클릭하면 해당 회원의 정보를 수정할 수 있는 페이지로 이동하도록 연결합니다.

yeonise committed:
User Controller에 회원 정보를 확인하고 수정하는 기능 추가

기존의 비밀번호가 확인된 경우에만 회원 정보를 수정할 수 있습니다.
---
현재 URL은 다음과 같습니다.
- 회원 가입
`GET` /users/join
- 회원 목록 조회
`GET` /users
- 회원 프로필 정보 보기
GET` /users/{userId}
- 회원 정보 수정 * 추가
`GET` /users/{userId}/update
`POST` /users/{userId}/update

yeonise committed:
이번 주 PR 마무리

이번 주 PR을 위해 마무리합니다.

yeonise committed:
fix : input name의 오타 수정

input name의 오타 때문에 글 내용이 전송되지 않는 오류가 있습니다.
content에서 contents로 수정합니다.

pbg0205 commented on README.md:
```diff
@@ -1,2 +1,119 @@
 # be-java-cafe
-마스터즈 2023 스프링 카페 
+
```
`README.md` 를 활용하는 목적은 프로젝트에서 사용 기술, 실행 방법, 요구사항에 관한 내용을 명시하는 목적으로 사용해요. 학습 계획의 경우는 별도의 파일로 분리하고 링크를 첨부해주는 방식으로 작성해주시면 작업하신 내용에 관한 내용을 명확히 파악할 수 있을 것 같습니다 😃

yeonise replied:
앞으로 `README.md`를 목적에 맞게 학습 계획이 아닌 프로젝트를 위한 용도로 사용해보겠습니다 🫡

pbg0205 commented on README.md:
```diff
@@ -1,2 +1,119 @@
+
+> 기본을 학습하여 작동 원리를 이해한다.
+
+어제 못한 계획을 우선순위에 두고 먼저 진행할 것 운동도 잊지 말자!
```
체력 관리 중요합니다 👍👍 ️

pbg0205 commented on README.md:
```diff
@@ -1,2 +1,119 @@
+2. 회원 프로필 조회 기능을 구현하면서 알게된 점 정리 (1H) ✔️
+   - `@PathVariable`를 통해 요청 파라미터를 받아서 사용할 수 있다.
+3. HTML의 중복 제거하기 (2H) ✔️
+   - Thymeleaf의 `Layout`을 사용하여 공통된 부분을 관리할 수 있다.
```
👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,49 @@
+@Controller
+public class ArticleController {
+
+    private final ArticleRepository articleRepository;
```
Repository 대신에 Service 레이어를 추가해보는건 어떨까요?
일반적인 layered architecture 의 경우는 `Controller - Service - Repository` 형태의 구조를 띄고 있습니다.
각 레이어의 역할과 책임에 대해 공부해보시는걸 추천드려요👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,49 @@
+        return "article/form";
+    }
+
+    @PostMapping("/articles/write")
```
`Post` 는 생성이라는 의미를 내포하고 있어서 `write` 가 없어도 생성한다는 의미를 던지는 것 같은데 @yeonise 의견은 어떠세요?

yeonise replied:
``` java
    @GetMapping("/articles/write")
    public String write() {
        return "article/form";
    }

    @PostMapping("/articles/write")
    public String write(@ModelAttribute Article article) {
        articleRepository.save(article);
        return "redirect:/";
    }
```

`@GetMapping /articles/write`에서 넘어가는 요청이라서 당연히 같은 url을 사용해야 한다고 생각했었는데,
생성한다는 행위는 Post method의 사용으로 명시되기 때문에 `Article` 이라는 자원을 가리키는 `/articles` 만으로도 충분한 것 같네요! 💡

yeonise replied:
혹시 이렇게 url을 변경하는 경우의 commit type은 `refactor` 에 해당할까요?
코드 자체를 변경하는 것은 아니지만, `fix` 는 버그 수정과 관련하여 사용된다고 하니 refactor에 더 가깝지 않나 싶은데 고민이 되네요 🤔

pbg0205 replied:
클라이언트 입장에서는 스펙이 변경되기 때문에 기능 변경의 영역에 해당해서  `feat: 글 생성 요청 URL 변경` 로 작성하면 될 것 같아요.
`refactor` 는 기능이 변경되지 않고 내부 로직이나 구조가 변경되는 것이고 `fix`의 경우는 기능이 비정상 동작을 의미하는데 명세가 변경되는 부분은 `feature` 로 사용하고 있습니다 😃

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,70 @@
+
+    private final UserRepository userRepository;
+
+    @Autowired
```
이 부분도 같이 제거해주세요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.List;
+
+@Controller
+@RequestMapping("/users")
```
공통화하는 @RequestMapping 사용 👍

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,70 @@
+        return "user/list";
+    }
+
+    @GetMapping("/join")
```
`sign-up` 단어는 어떠세요??

yeonise replied:
`sign-up` 과 `join` 중에 어떤 것을 사용할지 고민하다가 짧고 단순한 join을 선택했는데
그룹이나 조직에 가입한다는 뜻으로 넓게 사용되는 join보다,
웹 상에서 주로 쓰이면서 **회원 가입**이라는 구체적인 의미를 전달하는 `sign-up`이 더 적합한 것 같네요! 🧐

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,70 @@
+
+    @PostMapping("/join")
+    public String join(@ModelAttribute User user) { // ModelAttribute 이름 미지정 시 클래스 'User'의 첫 글자를 소문자로 바꾼 'user'로 자동 설정된다.
+        if (userRepository.isExists(user.getUserId())) {
+            throw new IllegalArgumentException("이미 사용 중인 아이디입니다");
+        }
+        userRepository.save(user);
```
입력 값에 대한 예외 검증하는 부분 좋네요 👍
조금 더 욕심 내보자면 커스텀 예외로 적용하면 좋을 것 같습니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/Article.java:
```diff
@@ -0,0 +1,58 @@
+    private String title;
+    private String writer;
+    private String contents;
+    private final LocalDateTime regTime;
```
약어는 가독성을 위해 지양해주세요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/Article.java:
```diff
@@ -0,0 +1,58 @@
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
```
setter 를 사용하기 보다는 의미있는 메서드명을 사용해보면 어떨까요?

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/Article.java:
```diff
@@ -0,0 +1,58 @@
+    public void setWriter(String writer) {
+        this.writer = writer;
+    }
+
```
사용하지 않는 코드는 제거해주세요~
setter 를 왜 지양할까요??

yeonise replied:
setter는 일반적으로 public으로 선언되어 클래스 외부에서 값을 변경할 수 있는데,
이렇게 여러 곳에서 직접 호출하여 사용하면 나중에 필드 값을 변경한 위치를 알아내기가 까다로워 오류를 찾거나 디버깅하는데 어려움이 있다고 하네요!
또 어떤 의도로 값을 변경했는지 알기 어려워서 위에서 남겨주신 코멘트처럼 **의도를 담은 메서드명**을 사용하는 것을 지향하는게 좋겠다는 생각이 들어요 💡

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,70 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
```
import 는 * 는 명확히 어떤 것을 imoprt 하는지 파악하기 어려워 지양합니다 
import 의 * 를 제거할 수 있는 intellij 설정 아래에 첨부해드릴게요~
추가적으로 intellij 에서는 auto import 설정을 지원해줘서 추가적으로 필요하시면 별도로 설정하셔도 될 것 같습니다 😃 
- https://pivox.tistory.com/46

yeonise replied:
설정을 수정했습니다! 🙆🏻 감사합니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
```diff
@@ -0,0 +1,60 @@
+    public User() {
+
+    }
+
```
사용하지 않는 빈 생성자는 제거해주세요

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,13 @@
+    Article save(Article article);
+    Article findById(Long id);
+    List<Article> findAll();
+    void update(Long id, Article article);
+    void delete(Long id);
+}
```
현재 단계에서는 사용하지 않는 메서드 입니다
필요한 경우에만 코드 추가하는 것을 추천합니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,43 @@
+        return store.containsKey(userId);
+    }
+
+    public void clearStore() {
+        store.clear();
+    }
+}
```
테스트를 위한 코드가 메인에 있는 것은 부자연스럽네요~ store 를 인스턴스 필드로 작성하는 방식은 어떻게 생각하세요?

yeonise replied:
`STORE`를 클래스 변수가 아닌 인스턴스 변수로 변경하면
clearStore() 대신, 테스트 코드에서 생성자를 통해 초기화하는 방법을 사용할 수 있을 것 같아요.
비즈니스 로직이 아닌 불필요한 코드는 지양하겠습니다 🙆🏻

``` java
    MemoryArticleRepository memoryArticleRepository = new MemoryArticleRepository();

    @AfterEach
    void afterEach() {
        memoryArticleRepository = new MemoryArticleRepository();
    }
```

pbg0205 replied:
`@BeforeEach` 를 쓰면 테스트 이전에 초기화하는 의미에서도 명확하고 중복도 제거할 수 있을 것 같은데 어떠세요??

```java
class MemoryArticleRepositoryTest {

    private MemoryArticleRepository memoryArticleRepository;

    @BeforeEach
    void setUp() {
        this.memoryArticleRepository = new MemoryArticleRepository();
    }
   ...
}
```

yeonise replied:
말씀하신 것처럼 '초기화'라는 의미에는 `@BeforEach` 가 더 명확한 것 같아요! 수정하겠습니다

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,49 @@
+    @GetMapping
+    public String listAllArticles(Model model) {
+        List<Article> articles = articleRepository.findAll();
+        model.addAttribute("articles", articles);
```
모델에 도메인을 추가하는 대신 DTO 로 전달하면 좋을 것 같아요~
DTO 를 사용하는 이유에 대해서 고민해보시면 좋을 것 같습니다.
아래 참고할만한 내용 첨부드릴게요!

- https://tecoble.techcourse.co.kr/post/2020-08-31-dto-vs-entity/

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @GetMapping("/articles/{id}")
+    public String getArticle(@PathVariable("id") Long id, Model model) {
```
@PathVariable 은 별도로 id 를 명시해주시 않으면 기본으로 필드명을 매핑하기 때문에 ("id") 부분은 제거해주셔도 될 것 같아요~

pbg0205 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,49 @@
+
+    private final ArticleRepository articleRepository;
+
+    @Autowired
```
생성자 방식을 사용할 떄는 생성자만 있으면 의존성 주입을 해주기 때문에 @Autowired 는 제거하셔도 될 것 같습니다
(추가적으로 생성자 방식은 스프링에서도 권장하는 방식이에요~ 빈 주입 방식에는 어떤 것들이 있고 생성자 방식의 장점에 대해 공부해보시면 좋을 것 같습니다(이 부분은 천천히 공부 해보시고 알려주세요ㅎㅎ))

pbg0205 commented on src/test/java/kr/codesqaud/cafe/repository/MemoryArticleRepositoryTest.java:
```diff
@@ -0,0 +1,57 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MemoryArticleRepositoryTest {
```
test code 👍👍

pbg0205 commented on src/test/java/kr/codesqaud/cafe/repository/MemoryArticleRepositoryTest.java:
```diff
@@ -0,0 +1,57 @@
+
+    @AfterEach
+    void afterEach() {
+        memoryArticleRepository.clearStore();
```
테스트를 위한 메인 코드는 제거해주세요~

yeonise replied:
네! `STORE` 를 인스턴스 변수로 변경하여, 생성자를 통해 초기화 하는 방식으로 수정했습니다

``` java
@AfterEach
    void afterEach() {
        memoryArticleRepository = new MemoryArticleRepository();
    }
```

다만 `static` 키워드가 있음에도 test를 통과하는 경우는 우연히 테스트가 `findAll() → save()` 순으로 진행되어 그런거겠죠?

``` java
public class MemoryArticleRepository implements ArticleRepository {

    private static final Map<Long, Article> STORE = new ConcurrentHashMap<>();

    ...

```

save()에서 Article을 저장하고 `STORE` 가 초기화되지 않은 상태에서 2개의 데이터를 추가로 더 삽입했기 때문에
` assertThat(allArticle.size()).isEqualTo(2);` 부분에서 테스트를 통과하지 못할 것으로 예상했는데 그렇지 않더라구요

``` java
@Test
    void findAll() {
        // given
        Article article1 = new Article("title1", "userId1", "content1");
        Article article2 = new Article("title2", "userId2", "content2");

        memoryArticleRepository.save(article1);
        memoryArticleRepository.save(article2);

        // when
        List<Article> allArticle = memoryArticleRepository.findAll();

        // then
        assertThat(allArticle.size()).isEqualTo(2);
        assertThat(allArticle).contains(article1, article2);
    }
```

pbg0205 replied:
저도 혹시 몰라서 JUnit 이 따로 지원하는게 있나 싶어서 확인봤더니 테스트 순서 때문인 것 같아요
정적 변수가 값은 들어가 있네요~
``` 
--- findall 테스트 전
STORE in constructor = {}

--- save 테스트 전
STORE in constructor = {1=Article{id=1, title='title1', writer='userId1', contents='content1', createdAt=2023-04-02T21:01:24.110338}, 2=Article{id=2, title='title2', writer='userId2', contents='content2', createdAt=2023-04-02T21:01:24.110397}}

```

yeonise replied:
순서 때문이군요 확인 감사합니다!

pbg0205 commented on src/main/java/kr/codesqaud/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,47 @@
+@Repository
+public class MemoryArticleRepository implements ArticleRepository {
+
+    private static final Map<Long, Article> store = new ConcurrentHashMap<>();
```
ConcurrentHashMap 👍(상수는 대문자로 적어주세요!)

pbg0205 commented on src/test/java/kr/codesqaud/cafe/repository/MemoryArticleRepositoryTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    void update() {
+    }
+
```
사용하지 않는 테스트 코드 제거해주세요~

pbg0205 commented on README.md:
```diff
@@ -1,2 +1,119 @@
+   `@EnableWebMvc`을 사용하면 `@Controller`를 붙인 컨트롤러를 위한 설정을 생성한다.  
+   이 마저도 SpringBoot에서는 `@SpringBootApplication`이 갖고 있는 `@EnableAutoConfiguration`을 통해 자동화하고 있다.
+   스프링에서는 `@Enable~`로 적용되는 인프라 bean에 대해 추가적인 설정을 할 수 있도록 `~Configurer`로 끝나는 인터페이스를 제공하고 있다.  
+   이를 구현한 클래스를 만들어 bean으로 등록하면 `@Enable~` 전용 어노테이션을 처리하는 단계에서 설정용 bean을 활용해 인프라 bean의 설정을 마무리한다.  
```
Java를 통해 빈을 등록하는 방법은 크게 두가지가 있는데 어떤 것이 있을까요?

yeonise replied:
찾아보니 생성한 Config 파일에 `@Configuration` 을 붙이고, 해당 파일 내에 `@Bean` 을 사용하여 등록하는 방법과
`@ComponentScan` 을 사용하여 해당 패키지를 포함한 하위 패키지에 있는 모든 `@Component` 가 붙은 클래스를 빈으로 등록하는 방법이 있다고 합니다

``` java
@ComponentScan(excludeFilters = { @Filter(type = FilterType.CUSTOM, classes = TypeExcludeFilter.class),
		@Filter(type = FilterType.CUSTOM, classes = AutoConfigurationExcludeFilter.class) })
public @interface SpringBootApplication {

    ...

```

CafeApplication에 붙인 `@SpringBootApplication` 도 내부적으로 `@ComponentScan` 을 사용하고 있어서 `@Component` 를 갖는 `@Repository` 와 `@Controller` 를 자동으로 빈으로 등록해주는 것 같아요!

pbg0205 replied:
잘 알고 계시네요 👍👍 `@ComponentScan` 의 기본 탐색 범위는 현재 위치한 클래스의 하위 클래스를 검색하기 때문에 검색 범위를 지정하고 싶은 경우에는 `basePackages` 로 별도로 할당할 수 있는 옵션이 있습니다. 한번 이런 옵션이 있구나 정도만 살펴보시면 좋을 것 같아요ㅎㅎ 기본기를 탄탄하게 학습하고 계신 것 같네요 😃

yeonise replied:
SpringBoot는 여러 가지로 간편하게 설정할 수 있는 옵션이 많네요! 한번 살펴보겠습니다 🧐

pbg0205 reviewed (changes requested):
안녕하세요 피아! 5주동안 함께 하게 된 리뷰어 쿠퍼입니다 😃
5주동안 같이 고민해보면서 좋은 코드 만들도록 노력해봐요🔥🔥

우선 README.md 에서 문의주신 내용에 관해서 답변드릴게요!

### 1. commit convention & 단위 에 관한 고민

- commit convention 은 아래 링크로 공유드릴게요. 링크 확인해보시고 알맞게 커밋 작업해주시면 될 것 같아요!
- commit 단위은 저의 경우에 커밋 컨벤션에 알맞게 구분하는 것 같아요.
- https://doublesprogramming.tistory.com/256
```
feat: 회원 등록 기능 개발

- User 도메인 추가 (User 생성시 검증 로직 추가)
- concurrentHashMap 을 통한 구현(InMemoryRepository)
- 회원 등록 시, 중복 사용자 커스텀 예외 추가 (AlreadyUserExistException)

```

예를 들어서 회원 등록에 대해서 커밋을 작성한다면 제목에 대표하는 제목을 적고, 하위에 특이 사항이나 중요한 내용에 대해 기입하는 방식으로 작업하고 있어요~ `1. commit type 에 따라서 구분, 2. 제목을 통해서 작업 내용 명시, 3. body에 고려한 상황이나 확인이 필요한 내용 첨부` 이와 같이 작성하는 편입니다. 저는 처음 commit 단위를 구분할 때는 주로 commit type 에 신경썼던 것 같아요. 한번 commit type 을 기준으로 일감을 구분해보시는 연습을 해보시면 좋을 것 같습니다! (e.g. feat: 회원 등록 기능 개발 -> test: 회원 테스트 작성 -> refactor: 미사용 import 제거) 


### 2. 예외에 관한 페이지 반환

예외 처리를 할 수 있는 방법은 @ControllerAdvice 를 사용하는 방법이 있어요~ 해당 컴포넌트는 예외를 잡아서 처리하는 컴포넌트인데 만약 @ControllerAdvice 를 적용하는데 어려움이 있으시면 예외 대신에 페이지 반환하는 방법도 있습니다~ 

```java
 @PostMapping("/join")
public String join(@ModelAttribute User user){
  if(userRepository.isExists(user.getUserId())){
    //throw new IllegalArgumentException("이미 사용 중인 아이디입니다");
    return "fail" // 실패 페이지 반환
  }
}
```

---

우선 코드 재정비하기 위해 `request change` 드렸습니다~ 작업해보시고 질문있으면 언제든 DM 주세요ㅎㅎ

yeonise committed:
Domain에서 사용하지 않는 setter 제거

Article과 User에서 사용하지 않는 setter와 생성자를 제거합니다.

yeonise committed:
사용하지 않는 test 코드 제거

- MemoryArticleRepositoryTest.java에서 사용하지 않는 빈 테스트 코드를 제거합니다.

yeonise committed:
생략 가능한 애노테이션 제거

- ArticleController와 UserController의 생성자에 붙은 @Autowired를 제거합니다.

yeonise committed:
Article 클래스에서 약어로 작성한 필드 수정

- 게시글을 등록한 시간을 저장하는 regTime을 createdAt으로 수정합니다.

yeonise committed:
Repository 내 테스트 코드를 위한 메서드 제거 및 테스트 코드 수정

- 상수는 대문자로 수정하였습니다.
- STORE를 클래스 변수에서 인스턴스 변수로 수정하였습니다.
- 테스트 코드에서 초기화가 필요한 경우, 인스턴스 변수의 생성자를 사용하여 초기화하는 방식으로 수정하였습니다.

yeonise committed:
생략 가능한 PathVariable의 name 삭제

- ArticleController와 UserController에서 생략 가능한 PathVariable의 name을 제거합니다.

pbg0205 reviewed (approved):
전체적으로 코드가 깔끔해졌네요 수고 많으셨습니다 😄 조금 더 욕심이 내보자면 `Service Layer` 도 추가하는 법도 고민해보시는 것도 좋을 것 같아요.
비록 간단한 CRUD 작업이지만 일반적으로 **Service Layer 는 우리 서비스의 비즈니스를 담당**, **Repository 는 DB 의 로직을 담당으로 구분하고 있어서** 서비스 게시글을 저장, 목록 조회한다는 측면에서 Service Layer 를 추가하고 Repository 를 통해 DB 에 저장하는 역할을 위임하는 구조로도 구분해서 작업하면 좋을 것 같다는 생각이 드네요ㅎㅎ
이 부분은 이후 스텝에서 고민해보시는 것으로 하고 추가적으로 개선해볼 포인트로 남겨두면 좋을 것 같습니다.
추가적으로 코멘트 남기신 부분에도 댓글 남겼으니 확인해보시고 다음 스텝에서 적용 여부에 관해 고민해보시면 좋을 것 같아요!
다시 한번 수고 많으셨고 다음 스텝에서 뵐게요 🖐

yeonise commented:
언급하신 **DTO**와 **Service Layer**를 더 알아보고 적용한 다음에 새로운 미션에 들어가려구요!
그리고 꼼꼼하게 남겨주신 코멘트들이 마치 응원처럼 힘이 되어요 감사합니다 🙇🏻 다음 주에 뵐게요 👋🏻
